core/src/main/kotlin/com/strumenta/kolasu/parsing/KolasuParser.kt

```diff
@@ -16,7 +16,40 @@ import java.util.*
 import kotlin.reflect.full.memberFunctions
 import kotlin.system.measureTimeMillis
 
-abstract class KolasuANTLRLexer : KolasuLexer<KolasuANTLRToken> {
+interface TokenFactory<T : KolasuToken> {
+    fun categoryOf(t: Token): TokenCategory = TokenCategory.PLAIN_TEXT
+
+    fun convertToken(t: Token): T
+
+    private fun convertToken(terminalNode: TerminalNode): T = convertToken(terminalNode.symbol)
+
+    fun extractTokens(result: ParsingResult<*>): LexingResult<T>? {
+        val antlrTerminals = mutableListOf<TerminalNode>()
+        fun extractTokensFromParseTree(pt: ParseTree?) {
+            if (pt is TerminalNode) {
+                antlrTerminals.add(pt)
+            } else if (pt != null) {
+                for (i in 0..pt.childCount) {
+                    extractTokensFromParseTree(pt.getChild(i))
+                }
+            }
+        }
+
+        val ptRoot = result.firstStage?.root
+        return if (ptRoot != null) {
+            extractTokensFromParseTree(ptRoot)
+            antlrTerminals.sortBy { it.symbol.tokenIndex }
+            val tokens = antlrTerminals.map { convertToken(it) }.toMutableList()
+            LexingResult(result.issues, tokens, result.code, result.firstStage.lexingTime)
+        } else null
+    }
+}
+
+class DefaultTokenFactory : TokenFactory<KolasuANTLRToken> {
```

Comment:
Could we call it ANTLRTokenFactory or DefaultANTLRTokenFactory?

Reply:
Sure!